doc/architecture/decisions/0002-implement-graphics-object-store.md

```diff
@@ -0,0 +1,423 @@
+operation completed.
+
+At the end of the operation, `transform` is called with the final
+transformation matrix, which then modifies the object in the object
+store.
+
+The 2D affine transformation functions are already implemented in
+`coordlist_ops.c`.
+
+### Actions
+
+When calling `exec`, the `action` argument selects the task to be performed.
+
+<table>
+<tr>
+<td> draw             </td><td> draw object in "normal" state </td>
+</tr>
+<tr>
+<td> draw_selected    </td><td> draw object in "selected" state </td>
+</tr>
+<tr>
+<td> draw_transformed </td><td> draw object in intermediate "transformed" state; <br>
+                                this allows to display the object dynamically, e.g. while it is being dragged </td>
+</tr>
+<tr>
+<tr>
+<td> update_bbox      </td><td> update the bounding box of the base object; <br>
+                                trigger recalculation of the bounding box </td>
+</tr>
+<tr>
+<td> get_is_clicked   </td><td> detect whether a click at coordinate (x,y) "hits" the object, for example to select it </td>
+</tr>
+<tr>
+<td> transform        </td><td> transforms the object, modifying the context in the <tt>gfx_store</tt> </td>
+</tr>
+<tr>
+<td> destroy          </td><td> destroys the object, deallocating memory </td>
+</tr>
+<tr>
+<td> clone            </td><td> deep-copies the object </td>
+</tr>
+</table>
+  
+
+### Drawing of object store
+
+Whenever objects need to be re-drawn, the `draw` function simply
+iterates through the object store and calls `object->do_action(ptr,
+draw_action)` for all objects. Objects with a bounding box not
+intersecting the dirty region (`gdk_cairo_get_clip_rectangle()`) are
+skipped.
+
+### Undo/redo system
+
+Each object that is added or transformed receives a new unique
+`id`. This `id` is delivered by a "factory" function and is a simple
+`guint32` that is incremented at each request.
+
+Maintaining a per-user (i.e. per device) undo system is complicated by
+the fact that an object can be "taken over" by another user by
+manipulating it. This is necessary to enable a truly collaborative
+workflow. At the same time, a user should only undo his/her own
+operations.
+
+To solve this conflict, I propose the following system:
+
+- each device (i.e. user) contains an `GList` with `undo` and `redo` records.
+
+- `GromitData` is extended with an undo object store (`undo_objects`).
+
+- the steps necessary to revert a user's operation are recorded in the
+  respective device's `undo` records.
+  
+- an undo operation executes the last undo operation according to the
+  `undo` record of the device, and moves it to the `redo` record.
+  
+- a redo operation does the reverse, and moves the record back to the
+  end of the `undo` records.
+  
+- any non-undo/redo operation clears the `redo` records.
+  
+The `undo` records contain the `id` of the object to move from
+`objects` to `undo_objects`, and the `id` of the object to move from
```

Comment:
Make it clear that `objects` is in GromitData and "global" please.

Reply:
this is written on line 35 already...